core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/ProgressiveUtils.scala

```diff
@@ -0,0 +1,38 @@
+package edu.uci.ics.texera.workflow.common
+
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, AttributeType}
+
+object ProgressiveUtils {
+
+  // boolean attribute to indicate insertion / retraction
+  // true  indicates insertion  (+)
+  // false indicates retraction (-)
+  val insertRetractFlagAttr = new Attribute("__internal_is_insertion", AttributeType.BOOLEAN)
+
+  def addInsertionFlag(tuple: Tuple): Tuple = {
+    assert(!tuple.getSchema.containsAttribute(insertRetractFlagAttr.getName))
+    Tuple.newBuilder.add(insertRetractFlagAttr, true).add(tuple).build
+  }
+
+  def addRetractionFlag(tuple: Tuple): Tuple = {
+    assert(!tuple.getSchema.containsAttribute(insertRetractFlagAttr.getName))
+    Tuple.newBuilder.add(insertRetractFlagAttr, false).add(tuple).build
+  }
```

Comment:
I did not check out `Tuple`'s builder logic, but does adding an attribute require making a copy of the same tuple?

Reply:
This is out of scope of this PR. But there will be no copy.